src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs

```diff
@@ -1,16 +1,121 @@
+            IConsensusManager consensusManager,
+            IDateTimeProvider dateTimeProvider,
+            Network network,
+            INodeLifetime nodeLifetime,
+            ILoggerFactory loggerFactory,
+            IInitialBlockDownloadState indState,
+            PoABlockDefinition blockDefinition)
+        {
+            this.consensusManager = consensusManager;
+            this.dateTimeProvider = dateTimeProvider;
+            this.network = network;
+            this.indState = indState;
+            this.blockDefinition = blockDefinition;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.cancellation = CancellationTokenSource.CreateLinkedTokenSource(new[] { nodeLifetime.ApplicationStopping });
+        }
+
+        /// <summary>Starts mining loop.</summary>
+        public void InitializeMining()
+        {
+            if (this.miningTask == null)
+            {
+                this.miningTask = this.CreateBlocksAsync();
+            }
+        }
+
+        private async Task CreateBlocksAsync()
+        {
+            while (!this.cancellation.IsCancellationRequested)
+            {
+                try
+                {
+                    // Don't mine in IBD.
+                    if (this.indState.IsInitialBlockDownload() && !MineDuringIBD)
+                    {
+                        int attemptDelayMs = 20_000;
+                        await Task.Delay(attemptDelayMs, this.cancellation.Token).ConfigureAwait(false);
+
+                        continue;
+                    }
+
+
+                    // TODO POA check if our timestamp and if not wait for our timestamp (wait with cancellation token)
+
+                    // TODO will set nonce and target to const. Check todo in PoABlockDefenition
+
+                    ChainedHeader tip = this.consensusManager.Tip;
+
+                    BlockTemplate blockTemplate = this.blockDefinition.Build(tip);
+
+                    // Timestamp should only greater than prev one.
+                    if (blockTemplate.Block.Header.Time <= tip.Header.Time)
+                        continue; // TODO POA Log
+
+                    // TODO sign the block with our signature
+
+                    ChainedHeader chainedHeader = await this.consensusManager.BlockMinedAsync(blockTemplate.Block).ConfigureAwait(false);
+
+                    if (chainedHeader == null)
+                    {
+                        this.logger.LogTrace("(-)[BLOCK_VALIDATION_ERROR]:false");
+                        continue;
+                    }
+
+                    // TODO POA Log mined block
+                }
+                catch (OperationCanceledException)
+                {
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
         {
+            this.cancellation.Cancel();
+            this.miningTask?.GetAwaiter().GetResult();
 
+            this.cancellation.Dispose();
```

Comment:
since we're doing a new Dispose method, can we maybe do dispose thew proper way?
Look here: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose
and for an example, look how it's done in the DNS project.

Reply:
But it's not override, would would we need overidable dispose(bool)?

Reply:
It's the pattern that's like that. this is so that you can call Dispose many times and the freeing of resources is properly managed + it doesn't throw exceptions. 
Look at this bit: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#dispose-and-disposeboolean

Reply:
Currently dispose method already supports calling it multiple times without a problem. 
What is the benefit of having 2 dispose (one with bool and one without) in this case?